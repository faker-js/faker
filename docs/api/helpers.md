# Helpers

<!-- This file is automatically generated. -->
<!-- Run 'pnpm run typedoc' to update -->

[[toc]]

::: v-pre

## Contextual Card

contextualCard

**Returns:** Object

```ts
faker.helpers.contextualCard(): Object
faker.helpers.contextualCard() // => {"name":"Katelynn","username":"Katelynn_Medhurs...
```

## Create Card

createCard

**Returns:** Object

```ts
faker.helpers.createCard(): Object
faker.helpers.createCard() // => {"name":"Lola Sporer","username":"Leatha_Stiede...
```

## Create Transaction

createTransaction

**Returns:** Object

```ts
faker.helpers.createTransaction(): Object
faker.helpers.createTransaction() // => {"amount":"548.81","date":"2012-02-01T23:00:00....
```

## Mustache

mustache

**Parameters**

- str?: string - N/A
- data?: Record<string, string | Function> - N/A

**Returns:** string

```ts
faker.helpers.mustache(str?: string, data?: Record<string, string | Function>): string
```

## Randomize

backward-compatibility

**Parameters**

- \<T\> - N/A
- array: readonly T[] = `...` - N/A

**Returns:** T

```ts
faker.helpers.randomize<T>(array: readonly T[] = ...): T
faker.helpers.randomize() // => "b"
```

## Regexp Style String Parse

parse string patterns in a similar way to RegExp

e.g. "#{3}test[1-5]" -> "###test4"

**Parameters**

- string: string = `''` - N/A

**Returns:** string

```ts
faker.helpers.regexpStyleStringParse(string: string = ''): string
faker.helpers.regexpStyleStringParse() // => ""
```

## Repeat String

String repeat helper, alternative to String.prototype.repeat.... See PR #382

**Parameters**

- string?: any - N/A
- num: number = `0` - N/A

**Returns:** string

```ts
faker.helpers.repeatString(string?: any, num: number = 0): string
```

## Replace Credit Card Symbols

replace symbols in a credit card schems including Luhn checksum

**Parameters**

- string: string = `'6453-####-####-####-###L'` - N/A
- symbol: string = `'#'` - N/A

**Returns:** string

```ts
faker.helpers.replaceCreditCardSymbols(string: string = '6453-####-####-####-###L', symbol: string = '#'): string
faker.helpers.replaceCreditCardSymbols() // => "6453-5578-6858-4663-4287"
```

## Replace Symbol With Number

Parses string for a symbol and replace it with a random number from 1-10

**Parameters**

- string: string = `''` - N/A
- symbol: string = `'#'` - defaults to `"#"`

**Returns:** string

```ts
faker.helpers.replaceSymbolWithNumber(string: string = '', symbol: string = '#'): string
faker.helpers.replaceSymbolWithNumber() // => ""
```

## Replace Symbols

Parses string for symbols (numbers or letters) and replaces them appropriately (# will be replaced with number,
? with letter and \* will be replaced with number or letter)

**Parameters**

- string: string = `''` - N/A

**Returns:** string

```ts
faker.helpers.replaceSymbols(string: string = ''): string
faker.helpers.replaceSymbols() // => ""
```

## Shuffle

Takes an array and randomizes it in place then returns it

Uses the modern version of the Fisherâ€“Yates algorithm

**Parameters**

- \<T\> - N/A
- o?: T[] - N/A

**Returns:** T[]

```ts
faker.helpers.shuffle<T>(o?: T[]): T[]
```

## Slugify

slugifies string

**Parameters**

- string: string = `''` - N/A

**Returns:** string

```ts
faker.helpers.slugify(string: string = ''): string
faker.helpers.slugify() // => ""
```

## Unique Array

Takes an array of strings or function that returns a string
and outputs a unique array of strings based on that source

**Parameters**

- \<T\> - N/A
- source?: T[] | Function - N/A
- length?: number - N/A

**Returns:** T[]

```ts
faker.helpers.uniqueArray<T>(source?: T[] | Function, length?: number): T[]
```

## User Card

userCard

**Returns:** Object

```ts
faker.helpers.userCard(): Object
faker.helpers.userCard() // => {"name":"Lola Sporer","username":"Leatha_Stiede...
```
