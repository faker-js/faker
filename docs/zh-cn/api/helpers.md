# 辅助函数

[[toc]]

## 数组中的随机元素

从提供的数组中返回一个随机值。

::: tip
| 参数 | 类型 | 默认值 |
| ----- | ----- | :---------------: |
| array | array | `["a", "b", "c"]` |
:::

```js
faker.helpers.randomize(); // a
faker.helpers.randomize(['bob', 'joe', 'tim']); // joe
```

## Slugify

将给定的字符串转为 slug。

::: tip
| 参数 | 类型 | 默认值 |
| ------ | ------ | :-----: |
| string | string | `""` |
:::

```js
faker.helpers.slugify(); // ""
faker.helpers.slugify('bob was here'); // bob-was-here
```

## 将符号替换为数字

将给定字符串中的特定符号替换为数字。

::: tip
| 参数 | 类型 | 默认值 |
| ------ | ------ | :-----: |
| string | string | `""` |
| symbol | string | `"#"` |
:::

```js
faker.helpers.replaceSymbolWithNumber(); // ""
faker.helpers.replaceSymbolWithNumber('bob###@example.com'); // bob790@example.com
faker.helpers.replaceSymbolWithNumber('bob????##???', '?'); // bob6269##849
```

## 替换符号

解析符号（数字或字母）的字符串并适当地替换它们。

- `#` 将被替换为数字；
- `?` 将被替换为字母；
- `*` 将被替换为数字或字母。

::: tip
| 参数 | 类型 | 默认值 |
| ------ | ------ | :-----: |
| format | string | `""` |
:::

```js
faker.helpers.replaceSymbols('#####'); // 98101
faker.helpers.replaceSymbols('???'); // ABC
faker.helpers.replaceSymbols('bob-###-42-??'); // bob-226-42-KB
```

## 打乱数组

随机打乱给定数组并返回。

::: tip
| 参数 | 类型 | 默认值 |
| ----- | ----- | :-----: |
| array | array | `[]` |
:::

```js
faker.helpers.shuffle(); // []
faker.helpers.shuffle(['a', 'b', 'c']); // ["c", "a", "b"]
```

## 模板

将字符串中的模板替换为给定的值。

::: tip
| 参数 | 类型 | 默认值 |
| ----- | ------ | :-----: |
| str | string | `n/a` |
| data | object | `n/a` |
:::

```js
faker.helpers.mustache(); // ""
faker.helpers.mustache('{{foo}} was {{baz}}', { foo: 'bar', baz: 42 }); // bar was 42
```

## 创建卡片

返回一个对象，但数据本身并不一致。

对于自洽的数据，请使用 [上下文卡片](/zh-cn/api/helpers.html#contextual-card)。

```js
faker.helpers.createCard();
```

```json
{
  "name": "Glen Hahn",
  "username": "Darrin_Champlin84",
  "email": "Benton_Swift30@hotmail.com",
  "address": {
    "streetA": "Dickinson Forest",
    "streetB": "63914 Eldora Forge",
    "streetC": "095 Bella Lodge Apt. 590",
    "streetD": "Apt. 980",
    "city": "East Allan",
    "state": "Nebraska",
    "country": "Liechtenstein",
    "zipcode": "08027",
    "geo": { "lat": "3.4797", "lng": "-123.6115" }
  },
  "phone": "(015) 568-3818 x649",
  "website": "lysanne.org",
  "company": {
    "name": "Funk - Nicolas",
    "catchPhrase": "Face to face dedicated moratorium",
    "bs": "turn-key benchmark web services"
  },
  "posts": [
    {
      "words": "enim molestias architecto",
      "sentence": "Beatae repellat deserunt eos.",
      "sentences": "Vero quae laudantium. Vel autem corrupti eligendi. Reiciendis itaque delectus deserunt ea error molestiae aperiam.",
      "paragraph": "Et sed nostrum placeat debitis maiores. Eos illum qui qui necessitatibus. Officiis a quisquam labore."
    },
    {
      "words": "qui dolor nihil",
      "sentence": "Occaecati asperiores rerum magni aspernatur eius id officiis.",
      "sentences": "Explicabo accusantium enim consequatur. Repellat placeat hic facere natus sint velit eligendi est distinctio.",
      "paragraph": "Fugiat maiores corrupti similique laboriosam enim culpa maiores velit. Distinctio consequatur illo commodi fuga quo repellendus. Nihil sequi dolor non. Nihil et blanditiis rerum cupiditate est et facilis aliquam."
    },
    {
      "words": "nesciunt iusto qui",
      "sentence": "Sapiente commodi facere laborum aut.",
      "sentences": "Molestias nemo fugiat itaque expedita est aspernatur praesentium explicabo repellat. Ea incidunt quia sint cupiditate saepe et tempora. Autem doloribus dolor eius omnis dolor. Eos laborum nesciunt iste rem placeat ut autem. Commodi error est non sapiente a.",
      "paragraph": "Eius maxime enim ut repellendus illum eum aut blanditiis. Quaerat qui omnis ab qui ipsum sint. Officiis iste neque ab qui dolor doloremque rerum quos sed."
    }
  ],
  "accountHistory": [
    {
      "amount": "251.84",
      "date": "2012-02-02T08:00:00.000Z",
      "business": "Breitenberg - Turcotte",
      "name": "Checking Account 0226",
      "type": "payment",
      "account": "66727594"
    },
    {
      "amount": "740.75",
      "date": "2012-02-02T08:00:00.000Z",
      "business": "Shields - Heller",
      "name": "Checking Account 3782",
      "type": "invoice",
      "account": "64889716"
    },
    {
      "amount": "378.68",
      "date": "2012-02-02T08:00:00.000Z",
      "business": "Dickens and Sons",
      "name": "Home Loan Account 1699",
      "type": "withdrawal",
      "account": "69892278"
    }
  ]
}
```

## 上下文卡片

返回名称、用户名和电子邮件自相关的对象。

```js
faker.helpers.contextualCard();
```

```json
{
  "name": "Joan",
  "username": "Joan39",
  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/ripplemdk/128.jpg",
  "email": "Joan39_Weimann93@gmail.com",
  "dob": "1973-09-26T20:12:25.191Z",
  "phone": "650-151-6699 x271",
  "address": {
    "street": "Boyle Points",
    "suite": "Apt. 093",
    "city": "Julioville",
    "zipcode": "79041",
    "geo": { "lat": "63.9355", "lng": "-150.2784" }
  },
  "website": "laurianne.info",
  "company": {
    "name": "Hudson and Sons",
    "catchPhrase": "Streamlined transitional firmware",
    "bs": "value-added incentivize communities"
  }
}
```

## 用户卡片

```js
faker.helpers.userCard();
```

```json
{
  "name": "Adriel Dach",
  "username": "Lawson.Rutherford83",
  "email": "Kassandra86@gmail.com",
  "address": {
    "street": "Gislason View",
    "suite": "Apt. 409",
    "city": "Tellyside",
    "zipcode": "00051",
    "geo": { "lat": "69.6104", "lng": "-109.3244" }
  },
  "phone": "1-685-232-7348",
  "website": "dakota.org",
  "company": {
    "name": "DuBuque Group",
    "catchPhrase": "Mandatory multi-state ability",
    "bs": "real-time grow methodologies"
  }
}
```

## 创建交易

返回一个表示交易的对象。

```js
faker.helpers.createTransaction();
```

```json
{
  "amount": "883.74",
  "date": "2012-02-02T08:00:00.000Z",
  "business": "Brakus LLC",
  "name": "Credit Card Account 9671",
  "type": "deposit",
  "account": "54758588"
}
```
