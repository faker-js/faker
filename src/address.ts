import type { Faker } from '.';
import type { Fake } from './fake';
import type { Helpers } from './helpers';

let f: Fake['fake'];

export class Address {
  readonly Helpers: Helpers;

  constructor(private readonly faker: Faker) {
    f = this.faker.fake;
    this.Helpers = this.faker.helpers;

    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Address.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }

    // TODO @Shinigami92 2022-01-13: Need better strategy
    // @ts-expect-error
    this.direction.schema = {
      description:
        'Generates a direction. Use optional useAbbr bool to return abbreviation',
      sampleResults: ['Northwest', 'South', 'SW', 'E'],
    };
    // @ts-expect-error
    this.cardinalDirection.schema = {
      description:
        'Generates a cardinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['North', 'South', 'E', 'W'],
    };
    // @ts-expect-error
    this.ordinalDirection.schema = {
      description:
        'Generates an ordinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['Northwest', 'Southeast', 'SW', 'NE'],
    };
  }

  /**
   * Generates random zipcode from format. If format is not specified, the
   * locale's zip format is used.
   *
   * @method faker.address.zipCode
   * @param format
   */
  zipCode(format?: string) {
    // if zip format is not specified, use the zip format defined for the locale
    if (typeof format === 'undefined') {
      const localeFormat = this.faker.definitions.address.postcode;
      if (typeof localeFormat === 'string') {
        format = localeFormat;
      } else {
        format = this.faker.random.arrayElement(localeFormat);
      }
    }
    return this.Helpers.replaceSymbols(format);
  }

  /**
   * Generates random zipcode from state abbreviation. If state abbreviation is
   * not specified, a random zip code is generated according to the locale's zip format.
   * Only works for locales with postcode_by_state definition. If a locale does not
   * have a postcode_by_state definition, a random zip code is generated according
   * to the locale's zip format.
   *
   * @method faker.address.zipCodeByState
   * @param state
   */
  zipCodeByState(state: string) {
    const zipRange = this.faker.definitions.address.postcode_by_state[state];
    if (zipRange) {
      return this.faker.datatype.number(zipRange);
    }
    return this.faker.address.zipCode();
  }

  /**
   * Generates a random localized city name. The format string can contain any
   * method provided by faker wrapped in `{{}}`, e.g. `{{name.firstName}}` in
   * order to build the city name.
   *
   * If no format string is provided one of the following is randomly used:
   *
   * * `{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}`
   * * `{{address.cityPrefix}} {{name.firstName}}`
   * * `{{name.firstName}}{{address.citySuffix}}`
   * * `{{name.lastName}}{{address.citySuffix}}`
   * * `{{address.cityName}}` when city name is available
   *
   * @method faker.address.city
   * @param format
   */
  city(format?: string | number) {
    const formats = [
      '{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}',
      '{{address.cityPrefix}} {{name.firstName}}',
      '{{name.firstName}}{{address.citySuffix}}',
      '{{name.lastName}}{{address.citySuffix}}',
    ];

    if (!format && this.faker.definitions.address.city_name) {
      formats.push('{{address.cityName}}');
    }

    if (typeof format !== 'number') {
      format = this.faker.datatype.number(formats.length - 1);
    }

    return f(formats[format]);
  }

  /**
   * Return a random localized city prefix
   *
   * @method faker.address.cityPrefix
   */
  cityPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_prefix
    );
  }

  /**
   * Return a random localized city suffix
   *
   * @method faker.address.citySuffix
   */
  citySuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_suffix
    );
  }

  /**
   * Returns a random city name
   *
   * @method faker.address.cityName
   */
  cityName(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_name
    );
  }

  /**
   * Returns a random localized street name
   *
   * @method faker.address.streetName
   */
  streetName(): string {
    let result: string;
    let suffix = this.faker.address.streetSuffix();
    if (suffix !== '') {
      suffix = ' ' + suffix;
    }

    switch (this.faker.datatype.number(1)) {
      case 0:
        result = this.faker.name.lastName() + suffix;
        break;
      case 1:
        result = this.faker.name.firstName() + suffix;
        break;
    }
    return result;
  }

  //
  // TODO: change all these methods that accept a boolean to instead accept an options hash.
  //

  /**
   * Returns a random localized street address
   *
   * @method faker.address.streetAddress
   * @param useFullAddress
   */
  streetAddress(useFullAddress: boolean = false): string {
    let address = '';
    switch (this.faker.datatype.number(2)) {
      case 0:
        address =
          this.Helpers.replaceSymbolWithNumber('#####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 1:
        address =
          this.Helpers.replaceSymbolWithNumber('####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 2:
        address =
          this.Helpers.replaceSymbolWithNumber('###') +
          ' ' +
          this.faker.address.streetName();
        break;
    }
    return useFullAddress
      ? address + ' ' + this.faker.address.secondaryAddress()
      : address;
  }

  /**
   * streetSuffix
   *
   * @method faker.address.streetSuffix
   */
  streetSuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_suffix
    );
  }

  /**
   * streetPrefix
   *
   * @method faker.address.streetPrefix
   */
  streetPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_prefix
    );
  }

  /**
   * secondaryAddress
   *
   * @method faker.address.secondaryAddress
   */
  secondaryAddress() {
    return this.Helpers.replaceSymbolWithNumber(
      this.faker.random.arrayElement(['Apt. ###', 'Suite ###'])
    );
  }

  /**
   * county
   *
   * @method faker.address.county
   */
  county(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.county
    );
  }

  /**
   * country
   *
   * @method faker.address.country
   */
  country(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.country
    );
  }

  /**
   * countryCode
   *
   * @method faker.address.countryCode
   * @param alphaCode default alpha-2
   */
  countryCode(alphaCode: string = 'alpha-2'): string {
    if (alphaCode === 'alpha-2') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code
      );
    }

    if (alphaCode === 'alpha-3') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code_alpha_3
      );
    }

    return this.faker.random.arrayElement(
      this.faker.definitions.address.country_code
    );
  }

  /**
   * state
   *
   * @method faker.address.state
   * @param useAbbr
   */
  // TODO christopher 2022-01-13: useAbbr not in use
  state(useAbbr?: boolean): string {
    return this.faker.random.arrayElement(this.faker.definitions.address.state);
  }

  /**
   * stateAbbr
   *
   * @method faker.address.stateAbbr
   */
  stateAbbr(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.state_abbr
    );
  }

  /**
   * latitude
   *
   * @method faker.address.latitude
   * @param max default is 90
   * @param min default is -90
   * @param precision default is 4
   */
  latitude(max: number = 90, min: number = -90, precision: number = 4): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   * longitude
   *
   * @method faker.address.longitude
   * @param max default is 180
   * @param min default is -180
   * @param precision default is 4
   */
  longitude(
    max: number = 180,
    min: number = -180,
    precision: number = 4
  ): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   *  direction
   *
   * @method faker.address.direction
   * @param useAbbr return direction abbreviation. defaults to false
   */
  direction(useAbbr: boolean = false) {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr
    );
  }

  /**
   * cardinal direction
   *
   * @method faker.address.cardinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  cardinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(0, 4)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(0, 4)
    );
  }

  /**
   * ordinal direction
   *
   * @method faker.address.ordinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  ordinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(4, 8)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(4, 8)
    );
  }

  nearbyGPSCoordinate(
    coordinate?: number[],
    radius?: number,
    isMetric?: boolean
  ): string[] {
    function randomFloat(min: number, max: number): number {
      return Math.random() * (max - min) + min;
    }
    function degreesToRadians(degrees: number): number {
      return degrees * (Math.PI / 180.0);
    }
    function radiansToDegrees(radians: number): number {
      return radians * (180.0 / Math.PI);
    }
    function kilometersToMiles(miles: number): number {
      return miles * 0.621371;
    }
    function coordinateWithOffset(
      coordinate: number[],
      bearing: number,
      distance: number,
      isMetric: boolean
    ): number[] {
      const R = 6378.137; // Radius of the Earth (http://nssdc.gsfc.nasa.gov/planetary/factsheet/earthfact.html)
      const d = isMetric ? distance : kilometersToMiles(distance); // Distance in km

      const lat1 = degreesToRadians(coordinate[0]); //Current lat point converted to radians
      const lon1 = degreesToRadians(coordinate[1]); //Current long point converted to radians

      const lat2 = Math.asin(
        Math.sin(lat1) * Math.cos(d / R) +
          Math.cos(lat1) * Math.sin(d / R) * Math.cos(bearing)
      );

      let lon2 =
        lon1 +
        Math.atan2(
          Math.sin(bearing) * Math.sin(d / R) * Math.cos(lat1),
          Math.cos(d / R) - Math.sin(lat1) * Math.sin(lat2)
        );

      // Keep longitude in range [-180, 180]
      if (lon2 > degreesToRadians(180)) {
        lon2 = lon2 - degreesToRadians(360);
      } else if (lon2 < degreesToRadians(-180)) {
        lon2 = lon2 + degreesToRadians(360);
      }

      return [radiansToDegrees(lat2), radiansToDegrees(lon2)];
    }

    // If there is no coordinate, the best we can do is return a random GPS coordinate.
    if (coordinate === undefined) {
      return [this.faker.address.latitude(), this.faker.address.longitude()];
    }
    radius ||= 10.0;
    isMetric ||= false;

    // TODO: implement either a gaussian/uniform distribution of points in circular region.
    // Possibly include param to function that allows user to choose between distributions.

    // This approach will likely result in a higher density of points near the center.
    const randomCoord = coordinateWithOffset(
      coordinate,
      degreesToRadians(Math.random() * 360.0),
      radius,
      isMetric
    );
    return [randomCoord[0].toFixed(4), randomCoord[1].toFixed(4)];
  }

  /**
   * Return a random time zone
   *
   * @method faker.address.timeZone
   */
  timeZone(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.time_zone
    );
  }
}
