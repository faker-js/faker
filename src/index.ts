import { Faker } from './faker';
import allLocales from './locales';

export type {
  AnimalDefinitions,
  ColorDefinitions,
  CommerceDefinitions,
  CommerceProductNameDefinitions,
  CompanyDefinitions,
  DatabaseDefinitions,
  DateDefinitions,
  DateEntryDefinition,
  FinanceCurrencyEntryDefinitions,
  FinanceDefinitions,
  HackerDefinitions,
  InternetDefinitions,
  LocaleDefinition,
  /** @deprecated Use LocationDefinitions instead */
  LocationDefinitions as AddressDefinitions,
  LocationDefinitions,
  LoremDefinitions,
  MusicDefinitions,
  /** @deprecated Use PersonDefinitions instead */
  PersonDefinitions as NameDefinitions,
  PersonDefinitions,
  /** @deprecated Use PersonTitleDefinitions instead */
  PersonTitleDefinitions as NameTitleDefinitions,
  PersonTitleDefinitions,
  PhoneNumberDefinitions,
  ScienceDefinitions,
  SystemDefinitions,
  SystemMimeTypeEntryDefinitions,
  VehicleDefinitions,
  WordDefinitions,
} from './definitions';
export { FakerError } from './errors/faker-error';
export type { FakerOptions, UsableLocale, UsedLocales } from './faker';
export type { AnimalModule } from './modules/animal';
export type {
  Casing,
  ColorFormat,
  ColorModule,
  CSSFunction,
  CSSSpace,
  NumberColorFormat,
  StringColorFormat,
} from './modules/color';
export type { CommerceModule } from './modules/commerce';
export type { CompanyModule } from './modules/company';
export type { DatabaseModule } from './modules/database';
export type { DatatypeModule } from './modules/datatype';
export type { DateModule } from './modules/date';
export type { FinanceModule } from './modules/finance';
export type { GitModule } from './modules/git';
export type { HackerModule } from './modules/hacker';
export type { HelpersModule } from './modules/helpers';
export type { ImageModule } from './modules/image';
export type { InternetModule } from './modules/internet';
export type {
  /** @deprecated Use LocationModule instead */
  LocationModule as AddressModule,
  LocationModule,
} from './modules/location';
export type { LoremModule } from './modules/lorem';
export type { MusicModule } from './modules/music';
export { Sex } from './modules/person';
export type {
  /** @deprecated Use PersonModule instead */
  PersonModule as NameModule,
  PersonModule,
  SexType,
} from './modules/person';
export type { PhoneModule } from './modules/phone';
export type { RandomModule } from './modules/random';
export type { ChemicalElement, ScienceModule, Unit } from './modules/science';
export type { StringModule } from './modules/string';
export type { SystemModule } from './modules/system';
export type { VehicleModule } from './modules/vehicle';
export type { WordModule } from './modules/word';
export { Faker };

// since we are requiring the top level of faker, load all locales by default
export const faker: Faker = new Faker({
  locales: allLocales,
});
