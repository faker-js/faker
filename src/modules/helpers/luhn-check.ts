/**
 * Checks that the given string passes the luhn algorithm.
 *
 * @param str The string to validate.
 */
export function luhnCheck(str: string): boolean {
  return luhnChecksum(str) === 0;
}

/**
 * Calculates the luhn check value for the given string.
 *
 * @param str The string to calculate the check value for.
 * May contain the `L` placeholder at the end.
 */
export function luhnCheckValue(str: string): number {
  const checksum = luhnChecksum(str.replace(/L?$/, '0'));
  return checksum === 0 ? 0 : 10 - checksum;
}

/**
 * Calculates the luhn checksum value for the given value.
 *
 * @param str The string to generate the checksum for.
 */
function luhnChecksum(str: string): number {
  str = str.replace(/[\s-]/g, '');
  let sum = 0;
  let alternate = false;
  for (let i = str.length - 1; i >= 0; i--) {
    let n = parseInt(str.substring(i, i + 1));
    if (alternate) {
      n *= 2;
      if (n > 9) {
        n = (n % 10) + 1;
      }
    }
    sum += n;
    alternate = !alternate;
  }
  return sum % 10;
}
