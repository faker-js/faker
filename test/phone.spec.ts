import { beforeEach, describe, expect, it } from 'vitest';
import { faker } from '../src';
import { luhnCheck } from '../src/modules/helpers/luhn-check';
import { seededTests } from './support/seededRuns';

const NON_SEEDED_BASED_RUN = 25;

describe('phone', () => {
  beforeEach(() => {
    faker.locale = 'en';
  });

  seededTests(faker, 'phone', (t) => {
    t.itEach('phoneFormats', 'imei');

    t.describeEach(
      'phoneNumber',
      'number'
    )((t) => {
      t.it('noArgs').it('format', '###-###-####');
    });

    t.describe('phoneNumberFormat', (t) => {
      t.it('noArgs');
      for (const index of [0, 1, 2, 3, 4]) {
        t.it(`with index = ${index}`, index);
      }
    });
  });

  describe(`random seeded tests for seed ${faker.seed()}`, () => {
    for (let i = 1; i <= NON_SEEDED_BASED_RUN; i++) {
      describe('phoneNumber()', () => {
        it('should return a random phoneNumber with a random format', () => {
          const phoneNumber = faker.phone.phoneNumber();

          expect(phoneNumber).toMatch(/\d/);
        });
      });

      describe('number()', () => {
        it('should return a random phoneNumber with a random format', () => {
          const phoneNumber = faker.phone.number();

          expect(phoneNumber).toMatch(/\d/);
        });
      });

      describe('phoneNumberFormat()', () => {
        it('should return phone number with proper US format (Array index)', () => {
          faker.locale = 'en';
          const phoneNumber = faker.phone.phoneNumberFormat(1);
          expect(phoneNumber).toMatch(/\([2-9]\d\d\) [2-9]\d\d-\d\d\d\d/);
        });

        it('should return phone number with proper CA format (Array index)', () => {
          faker.locale = 'en_CA';
          const phoneNumber = faker.phone.phoneNumberFormat(1);
          expect(phoneNumber).toMatch(/\([2-9]\d\d\)[2-9]\d\d-\d\d\d\d/);
        });

        it('should return phone number with proper PL format (Array index)', () => {
          faker.locale = 'pl';
          const phoneNumber = faker.phone.phoneNumberFormat(1);
          expect(phoneNumber).toMatch(/13-\d{3}-\d{2}-\d{2}/);
        });
      });

      describe('phoneFormats()', () => {
        it('should return random phone number format', () => {
          const phoneFormat = faker.phone.phoneFormats();
          expect(faker.definitions.phone_number.formats).contain(phoneFormat);
        });
      });

      describe('imei()', () => {
        it('should return a string', () => {
          const imei = faker.phone.imei();
          expect(imei).toBeTypeOf('string');
        });

        it('should have a length of 18', () => {
          const imei = faker.phone.imei();
          expect(imei).toHaveLength(18);
        });

        it('should be Luhn-valid', () => {
          const imei = faker.phone.imei();
          expect(imei).toSatisfy(luhnCheck);
        });
      });
    }
  });
});
