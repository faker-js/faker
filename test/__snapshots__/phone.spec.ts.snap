// Vitest Snapshot v1

exports[`phone > 42 > imei 1`] = `"37-917755-141004-5"`;

exports[`phone > 42 > number > format 1`] = `"379-177-5514"`;

exports[`phone > 42 > number > noArgs 1`] = `"891.775.5141"`;

exports[`phone > 42 > phoneFormats 1`] = `"!##.!##.####"`;

exports[`phone > 42 > phoneNumber > format 1`] = `"379-177-5514"`;

exports[`phone > 42 > phoneNumber > noArgs 1`] = `"891.775.5141"`;

exports[`phone > 42 > phoneNumberFormat > noArgs 1`] = `"479-377-5514"`;

exports[`phone > 42 > phoneNumberFormat > with index = 0 1`] = `"479-377-5514"`;

exports[`phone > 42 > phoneNumberFormat > with index = 1 1`] = `"(479) 377-5514"`;

exports[`phone > 42 > phoneNumberFormat > with index = 2 1`] = `"1-479-377-5514"`;

exports[`phone > 42 > phoneNumberFormat > with index = 3 1`] = `"479.377.5514"`;

exports[`phone > 42 > phoneNumberFormat > with index = 4 1`] = `"479-377-5514"`;

exports[`phone > 1211 > imei 1`] = `"94-872190-616274-4"`;

exports[`phone > 1211 > number > format 1`] = `"948-721-9061"`;

exports[`phone > 1211 > number > noArgs 1`] = `"1-587-319-0616 x27431"`;

exports[`phone > 1211 > phoneFormats 1`] = `"1-!##-!##-#### x#####"`;

exports[`phone > 1211 > phoneNumber > format 1`] = `"948-721-9061"`;

exports[`phone > 1211 > phoneNumber > noArgs 1`] = `"1-587-319-0616 x27431"`;

exports[`phone > 1211 > phoneNumberFormat > noArgs 1`] = `"948-821-9061"`;

exports[`phone > 1211 > phoneNumberFormat > with index = 0 1`] = `"948-821-9061"`;

exports[`phone > 1211 > phoneNumberFormat > with index = 1 1`] = `"(948) 821-9061"`;

exports[`phone > 1211 > phoneNumberFormat > with index = 2 1`] = `"1-948-821-9061"`;

exports[`phone > 1211 > phoneNumberFormat > with index = 3 1`] = `"948.821.9061"`;

exports[`phone > 1211 > phoneNumberFormat > with index = 4 1`] = `"948-821-9061"`;

exports[`phone > 1337 > imei 1`] = `"25-122540-325523-4"`;

exports[`phone > 1337 > number > format 1`] = `"251-225-4032"`;

exports[`phone > 1337 > number > noArgs 1`] = `"(612) 454-0325"`;

exports[`phone > 1337 > phoneFormats 1`] = `"(!##) !##-####"`;

exports[`phone > 1337 > phoneNumber > format 1`] = `"251-225-4032"`;

exports[`phone > 1337 > phoneNumber > noArgs 1`] = `"(612) 454-0325"`;

exports[`phone > 1337 > phoneNumberFormat > noArgs 1`] = `"451-325-4032"`;

exports[`phone > 1337 > phoneNumberFormat > with index = 0 1`] = `"451-325-4032"`;

exports[`phone > 1337 > phoneNumberFormat > with index = 1 1`] = `"(451) 325-4032"`;

exports[`phone > 1337 > phoneNumberFormat > with index = 2 1`] = `"1-451-325-4032"`;

exports[`phone > 1337 > phoneNumberFormat > with index = 3 1`] = `"451.325.4032"`;

exports[`phone > 1337 > phoneNumberFormat > with index = 4 1`] = `"451-325-4032"`;
